(towards #1010) Use the PSyIR Fortran backend for LFRic loops and anything inside them

sergisiso opened the pull request.

sergisiso commented:
@arporter This is ready for a first review. The Codecov fail seems to be because I reduced the total number of lines and this changes the final percentage, but I believe everything touched by this PR is covered (minus some lines in f2pygen where I added the "pragma: no cover"). I also triggered the integration tests.

arporter commented on src/psyclone/dynamo0p3.py:
```diff
@@ -2335,7 +2335,13 @@ def __init__(self, invoke, unique_psy_vars):
+            # call will not be able to use the ig_kernels to find its
+            # information, therefore we also store a direct link in the
+            # kernel
+            self._ig_kernels[id(call)] = intergrid
```
Doing a quick grep, I see that (after your changes in this PR) we only ever loop over the `values` in the `_ig_kernels` dict and therefore it could perhaps just be a list instead. Perhaps add this as a task to #2503?

sergisiso replied:
Even better, I got rid of the cached dict/list altogether to avoid the double reference with self._schedule.walk(CodedKern)._itergrid_ref. Now the property looks for them dynamically, giving the right answer after tree schedule updates.

arporter replied:
Woohoo!

arporter commented on src/psyclone/dynamo0p3.py:
```diff
@@ -5522,21 +5527,6 @@ def dofs(self):
         and remove the need for this property (#279). '''
         return self._dofs
 
-    def raw_arg_list(self):
```
More whoops :-)

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -719,7 +719,8 @@ def __init__(self, name, KernFactory, BuiltInFactory, alg_calls=None,
         if reserved_names:
             for reserved in reserved_names:
-                self.symbol_table.add(Symbol(reserved))
+                self.symbol_table.add(Symbol(reserved),
```
Is this change related to the fact that you removed one of the entries in reserved names in a file I reviewed earlier?

sergisiso replied:
I am not sure, I could revert this change without tests failing.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -1531,18 +1532,30 @@ def lower_to_language_level(self):
+            except KeyError as err:
+                raise GenerationError(
+                    f"Cannot generate this kernel call to '{self.name}' "
+                    f"because it is marked as module-inline but no such "
```
"module-inlined" and (below) "exists".

arporter commented on src/psyclone/psyir/nodes/loop.py:
```diff
@@ -437,40 +437,53 @@ def gen_code(self, parent):
+                if kernel.modified:
+                    kernel.rename_and_write()
+
+        # Use the Fortran Backend from this point 
```
pylint complains about some trailing whitespace in this file.

sergisiso replied:
Fixed

arporter commented on src/psyclone/psyir/nodes/loop.py:
```diff
@@ -437,40 +437,53 @@ def gen_code(self, parent):
+        # uses the lowering+backend), these are:
+        # - Declaring the loop variable symbols
+        for loop in self.walk(Loop):
+            if loop._variable is None:
```
Could this be `loop.variable` to save the pylint error?

sergisiso replied:
Unfortunately no because the property does validations, I disabled the pylint for now as this should go with 1010

arporter commented on src/psyclone/tests/domain/lfric/lfric_domain_kernels_test.py:
```diff
@@ -297,8 +297,8 @@ def test_psy_gen_domain_kernel(dist_mem, tmpdir, fortran_writer):
     assert (expected + "      !\n"
-            "      !\n"
-            "      call testkern_domain_code(nlayers, ncell_2d_no_halos, b, "
+            "      do dummy = 1, 1, 1\n"
```
As queried earlier, can't we just remove this loop altogether?

sergisiso replied:
I now got rid of this loops as in the original.

arporter commented on src/psyclone/tests/domain/lfric/lfric_loop_test.py:
```diff
@@ -261,39 +260,23 @@ def test_lower_to_language_normal_loop():
-    fuses the 'loops' to verify that the kernels are all still in the right
-    order.
+def test_lower_to_language_null_loop():
+    ''' Tests that we can call lower_to_language_level on a NULL LFRicLoop.
```
Strictly speaking it is called a 'domain' LFRicLoop so I suggest "'domain'/NULL"

sergisiso replied:
Done

arporter commented on src/psyclone/tests/domain/lfric/lfric_loop_test.py:
```diff
@@ -261,39 +260,23 @@ def test_lower_to_language_normal_loop():
     _, invoke = get_invoke("25.1_kern_two_domain.f90", TEST_API, idx=0)
-    # Domain loops cannot be fused with the transformation, so manually
-    # move the two kernels into one domain loop. First detach the second
+    # For this test we want to put 2 kernels inside the same null loop
```
This comment needs updating because we no longer manually move the two kernels into a single loop. In fact, I'm not quite sure what is going on so please add comments :-)

arporter commented on src/psyclone/tests/domain/lfric/lfric_loop_test.py:
```diff
@@ -261,39 +260,23 @@ def test_lower_to_language_normal_loop():
-    assert sched.children[0].children[3].name == "f1_data"
-    assert sched.children[1].children[2].name == "b"
-    assert sched.children[1].children[3].name == "f2_data"
+    assert not isinstance(sched.children[1], LFRicLoop)
```
Should it just be a `Loop` now? Please add a check.

arporter replied:
Please could you also check that the final 'body' of the loop is as expected (just check the `debug_string()` perhaps?).

sergisiso replied:
Aha this loop was testing a domain loop with 2 statements inside. 

In fact this is the reason why I may had the dummy loop in my previous version, PSyIRGen only accepts one node. For now I raise and error for domain loops with multiple statements as they can not be fused anyway, and add a new test for this. But then everything is part of PSyIR the lowering can add multiple statements to its parent lowered schedule.

arporter commented on src/psyclone/tests/gocean1p0_test.py:
```diff
@@ -1083,35 +1083,6 @@ def test_find_grid_access(monkeypatch):
     assert arg is None
 
 
```
pylint says there are a lot of unused imports in this file.

sergisiso replied:
Done

arporter reviewed (changes requested):
Exciting stuff Sergi. Mostly a big improvement but you'll see I have a couple of questions. The main one being about LFRic 'domain' kernels (that don't have a loop over cells). I've fired off the integration tests and they're running now.

sergisiso commented on src/psyclone/domain/lfric/arg_ordering.py:
```diff
@@ -100,6 +92,25 @@ def __init__(self, kern):
+        elif self._kern and self._kern.ancestor(psyGen.InvokeSchedule):
+            return self._kern.ancestor(psyGen.InvokeSchedule).symbol_table
+        else:
+            return LFRicSymbolTable()
```
We could cache it by assigning it to _forced_symtab, but then it stays cached even when we connect this arg_ordering to a kernel and we do this when testing. The behaviour of not preserving the symbol table if we have not forced one or has an invoke is what kept existing tests correct, so I would prefer not changing the behaviour until we address 2503. But I have updated the docstring to make it clearer what it does.

sergisiso commented on src/psyclone/domain/lfric/kern_call_arg_list.py:
```diff
@@ -933,11 +934,14 @@ def cell_ref_name(self, var_accesses=None):
         if self._kern.is_coloured():
-            colour_sym = self._symtab.find_or_create_integer_symbol(
-                "colour", tag="colours_loop_idx")
+            colour_sym = self._symtab.find_or_create_tag(
```
This happens when the node is not connected to any invoke scope, which is where the symbols should go. I would say that if we want to preserve the symbols without the scope the tests need to provide the desired forced_symtab. Hopefully this all can be improved with 2503.

sergisiso commented on src/psyclone/domain/lfric/kern_call_arg_list.py:
```diff
@@ -933,11 +934,14 @@ def cell_ref_name(self, var_accesses=None):
         '''
-        cell_sym = self._symtab.find_or_create_integer_symbol(
-            "cell", tag="cell_loop_idx")
+        cell_sym = self._symtab.find_or_create_tag(
```
Reverted to KIND=i_def as discussed (also the colour variable below)

sergisiso commented on src/psyclone/domain/lfric/lfric_builtins.py:
```diff
@@ -458,8 +444,9 @@ def get_dof_loop_index_symbol(self):
         # constructor.
-        return table.find_or_create_integer_symbol(
-            "df", tag="dof_loop_idx")
+        return table.find_or_create_tag(
```
Reverted to i_def

sergisiso commented on src/psyclone/domain/lfric/lfric_invoke.py:
```diff
@@ -80,7 +80,6 @@ def __init__(self, alg_invocation, idx, invokes):
         reserved_names_list = []
         const = LFRicConstants()
         reserved_names_list.extend(const.STENCIL_MAPPING.values())
-        reserved_names_list.extend(const.VALID_STENCIL_DIRECTIONS)
```
Now these are handled by the PSyIR (specifically in the kernel call lowering), and to always refer to the same symbol it uses tags.

I believe the reserved_name_list is because some these are/were handled as strings produced by gen_code and we don't want to have symbols and gen_code strings with the same name, but we can get rid of all this when everything is in PSyIR.

sergisiso commented on src/psyclone/domain/lfric/lfric_kern.py:
```diff
@@ -105,6 +105,7 @@ def __init__(self):
         self._qr_rules = OrderedDict()
         self._cma_operation = None
         self._is_intergrid = False  # Whether this is an inter-grid kernel
+        self._intergrid_ref = None  # Reference to this kernel inter-grid
```
Done, but to achieve this I had to move its initialisation from DynMesh to this class setup(), which I think is a good change since we can use 'self' instead external private methods for this functionality.

arporter replied:
Great :-)

sergisiso commented on src/psyclone/domain/lfric/lfric_kern.py:
```diff
@@ -400,12 +401,11 @@ def colourmap(self):
         if self._is_intergrid:
             invoke = sched.invoke
-            if id(self) not in invoke.meshes.intergrid_kernels:
+            if self._intergrid_ref is None:
```
Yes, this pattern was repeated a few times and I got rid of all of them.

sergisiso commented on src/psyclone/domain/lfric/lfric_kern.py:
```diff
@@ -474,13 +472,11 @@ def ncolours_var(self):
         if self._is_intergrid:
-            invoke = self.ancestor(InvokeSchedule).invoke
-            if id(self) not in invoke.meshes.intergrid_kernels:
+            if self._intergrid_ref is None:
```
Done

sergisiso commented on src/psyclone/domain/lfric/lfric_loop.py:
```diff
@@ -98,8 +99,17 @@ def __init__(self, loop_type="", **kwargs):
+                    tag, root_name=suggested_name, symbol_type=DataSymbol,
+                    datatype=INTEGER_TYPE)
+            else:
+                # TODO: Shouldn't all these all be INTEGER_TYPE instead of
```
Reverted to i_def

sergisiso commented on src/psyclone/dynamo0p3.py:
```diff
@@ -2469,8 +2475,7 @@ def _colourmap_init(self):
                     base_name, 1, ScalarType.Intrinsic.INTEGER, tag=base_name)
             # Add these symbols into the dictionary entry for this
```
Done

sergisiso commented on src/psyclone/domain/lfric/lfric_kern.py:
```diff
@@ -965,6 +887,67 @@ def _validate_kernel_code_arg(self, kern_code_arg, interface_arg,
 
+    def validate_global_constraints(self):
+        '''
+        Generates LFRic (Dynamo 0.3) specific PSy layer code for a call
```
Ups, updated!

sergisiso commented on src/psyclone/domain/lfric/lfric_kern.py:
```diff
@@ -965,6 +887,67 @@ def _validate_kernel_code_arg(self, kern_code_arg, interface_arg,
+        '''
+        Generates LFRic (Dynamo 0.3) specific PSy layer code for a call
+        to this user-supplied LFRic kernel.
+        :raises GenerationError: if this kernel does not have a supported \
```
Done

sergisiso commented on src/psyclone/domain/lfric/lfric_kern.py:
```diff
@@ -965,6 +887,67 @@ def _validate_kernel_code_arg(self, kern_code_arg, interface_arg,
+        # Get configuration for valid argument kinds
+        api_config = Config.get().api_conf("dynamo0.3")
+
+        from psyclone.domain.lfric import LFRicLoop
```
Done (was not possible to move up)

sergisiso commented on src/psyclone/domain/lfric/lfric_kern.py:
```diff
@@ -965,6 +887,67 @@ def _validate_kernel_code_arg(self, kern_code_arg, interface_arg,
+        if op_args:
+            # It does. We must check that our parent loop does not
+            # go beyond the L1 halo.
+            if parent_loop.upper_bound_name == "cell_halo" and \
```
Done

sergisiso commented on src/psyclone/domain/lfric/lfric_kern.py:
```diff
@@ -965,6 +887,67 @@ def _validate_kernel_code_arg(self, kern_code_arg, interface_arg,
+                except FieldNotFoundError:
+                    arg = None
+                if arg:
+                    raise GenerationError(f"Kernel '{self._name}' has an "
```
Done

sergisiso commented on src/psyclone/domain/lfric/lfric_loop.py:
```diff
@@ -128,30 +138,45 @@ def lower_to_language_level(self):
-        for child in all_children_reverse:
-            parent.children.insert(pos, child)
-        return parent
+            # This is a domain loop, first check that there isn't any
```
Done

sergisiso commented on src/psyclone/domain/lfric/lfric_loop.py:
```diff
@@ -128,30 +138,45 @@ def lower_to_language_level(self):
+            self.replace_with(loop)
+        else:
+            # If loop_type is "null" we no need for a loop at all. So we set
+            # the bounds to 1 Remove the lower the children
```
yeah we can, done!

sergisiso commented on src/psyclone/domain/lfric/lfric_loop.py:
```diff
@@ -838,13 +863,7 @@ def gen_code(self, parent):
-            # a loop - we go on down to the children instead.
-            for child in self.loop_body.children:
-                child.gen_code(parent)
+        super().gen_code(parent)
```
gen_code calls PSyIRGen(self.lower_to_language_level) which already does the right thing for null loops.

The reason I cannot get rid of the remainer of this method yet is that it modifies siblings (potentially adding a haloex) but the lowered psyir starts here, so I have a boundary where modifications of the lowering will not modify parent/siblings. The null loop modification happens inside the loop, so is now managed by the lowering.

We can get rid of this gen_code at the next step which is to make the whole InvokeSchedule use lowering+backend.

I added a TODO to clarify all this.

arporter replied:
Thanks :-)

sergisiso commented:
@arporter This is ready for another review

arporter commented on src/psyclone/domain/lfric/lfric_kern.py:
```diff
@@ -399,12 +412,7 @@ def colourmap(self):
-                raise InternalError(
-                    f"Colourmap information for kernel '{self.name}' has "
-                    f"not yet been initialised")
+        if self.is_intergrid:
```
Please also update the docstring description of when this will raise an InternalError.

sergisiso replied:
Done

arporter commented on src/psyclone/domain/lfric/lfric_kern.py:
```diff
@@ -435,11 +443,7 @@ def last_cell_all_colours_symbol(self):
-                raise InternalError(
-                    f"Colourmap information for kernel '{self.name}' has "
-                    f"not yet been initialised")
+        if self.is_intergrid:
```
Please update the 'raises' bit of the docstring.

sergisiso replied:
Done

arporter commented on src/psyclone/domain/lfric/lfric_kern.py:
```diff
@@ -471,11 +475,7 @@ def ncolours_var(self):
-                raise InternalError(
-                    f"Colourmap information for kernel '{self.name}' has "
-                    f"not yet been initialised")
+        if self.is_intergrid:
```
'raises' docstring update please.

sergisiso replied:
Done

arporter commented on src/psyclone/domain/lfric/lfric_kern.py:
```diff
@@ -888,14 +888,15 @@ def _validate_kernel_code_arg(self, kern_code_arg, interface_arg,
-        Generates LFRic (Dynamo 0.3) specific PSy layer code for a call
-        to this user-supplied LFRic kernel.
-        :raises GenerationError: if this kernel does not have a supported \
+        Perform validation checks for any global constraints (that requiere the
```
"require"

sergisiso replied:
Fixed

arporter commented on src/psyclone/dynamo0p3.py:
```diff
@@ -2342,7 +2320,7 @@ def declarations(self, parent):
 
-        if not self._ig_kernels and (self._needs_colourmap or
+        if not self.intergrid_kernels and (self._needs_colourmap or
                                      self._needs_colourmap_halo):
```
Indentation needs fixing?

sergisiso replied:
Done

arporter commented on src/psyclone/dynamo0p3.py:
```diff
@@ -2545,12 +2523,15 @@ def initialise(self, parent):
-        :rtype: :py:class:`collections.OrderedDict`
         '''
-        return self._ig_kernels
+        :returns: A list of intergrid mesh objects used.
```
"A list of objects describing the intergrid kernels used in this invoke."

arporter commented on src/psyclone/dynamo0p3.py:
```diff
@@ -2545,12 +2523,15 @@ def initialise(self, parent):
         '''
-        return self._ig_kernels
+        :returns: A list of intergrid mesh objects used.
+        :rtype: List[:py:class:`psyclone.dynamo3p0.DynInterGrid`]
```
`list[...]` please.

sergisiso replied:
Done

arporter commented on src/psyclone/domain/lfric/lfric_loop.py:
```diff
@@ -160,23 +151,23 @@ def lower_to_language_level(self):
+            # TODO #1010: This restriction can be removed when also lowering the
+            # parent InvokeSchedule
+            if len(self.loop_body.children) > 1:
+                raise GenerationError(
```
Strictly this should be a `NotImplementedError` unless that causes problems elsewhere?

sergisiso commented on src/psyclone/domain/lfric/lfric_loop.py:
```diff
@@ -160,23 +151,23 @@ def lower_to_language_level(self):
-            dummy = self.scope.symbol_table.new_symbol(
-                "dummy", symbol_type=DataSymbol, datatype=INTEGER_TYPE)
-            self._variable = dummy
+            # If loop_type is "null" we not need a loop at all, just the kernel
```
Ups, fixed

sergisiso commented on src/psyclone/tests/domain/lfric/lfric_domain_kernels_test.py:
```diff
@@ -448,6 +442,7 @@ def test_domain_plus_cma_kernels(dist_mem, tmpdir):
     assert "mesh => f1_proxy%vspace%get_mesh()" in gen_code
     assert "ncell_2d = mesh%get_ncells_2d()" in gen_code
     assert "ncell_2d_no_halos = mesh%get_last_edge_cell()" in gen_code
+    print(gen_code)
```
Done

sergisiso commented:
@arporter This is ready for next review

arporter reviewed (changes requested):
All requested changes have been made.
The newly-automated linting is highlighting some issues - please could you fix. Once that's done this can go on.

sergisiso commented:
@arporter I have now fixed the remaining formatting errors. I also removed an uncovered check in the stub generator, the reason is that now intergrid kernels can not be created outside an InvokeSchedule scope, as they need the path to the mesh symbols, so the test at src/psyclone/tests/dynamo0p3_multigrid_test.py line 772 already fails to create the necessary stub intergrid kernel before creating the stub itself.

This is ready to review, CI-permitting.

arporter reviewed (approved):
All requested changes have been made.
Ref. guide still builds fine.
We'll need to update the docs as we move to adopting the PSyIR backend for LFRic but that can be done when we're a bit further down the line.
Integration tests are running.

arporter commented:
Just waiting on the integration tests now...

arporter commented:
Integration tests were green. Ready to go once CI complete.
